substrate-node/pallets/pallet-smart-contract/src/billing.rs

```diff
@@ -386,6 +392,26 @@ impl<T: Config> Pallet<T> {
                     diff
                 );
                 if diff >= T::GracePeriod::get() {
+                    // Ensure associated node contracts have no reserved balance.
+                    // If none, proceed to move the contract to the 'deleted' state.
+                    // Otherwise, wait.
+                    let can_be_deleted = match contract.contract_type {
+                        types::ContractData::RentContract(_) => {
+                            let active_node_contracts = ActiveNodeContracts::<T>::get(contract.get_node_id());
+                            !active_node_contracts.iter().any(|id| {
+                                ContractPaymentState::<T>::get(id).map_or(false, |s| s.has_reserve())
+                            })
+                        }
+                        _ => true,
+                    };
```

Comment:
Can we have more explicit naming here and insert `if` inside `match` ?
Like:
~~~rust
                    match contract.contract_type {
                        types::ContractData::RentContract(_) => {
                            let active_node_contracts = ActiveNodeContracts::<T>::get(contract.get_node_id());
                            let rent_has_node_contracts_with_reserve = 
                            active_node_contracts.iter().any(|id| {
                                ContractPaymentState::<T>::get(id).map_or(false, |s| s.has_reserve())
                            })

                            if rent_has_node_contracts_with_reserve {
                                log::debug!(
                                    "Grace period expired, but one or more associated node contracts have held user funds. \
                                    Rent contract deletion will be delayed until funds are distributed to beneficiaries."
                                );
                                return Ok(().into());
                            }
                        }
                        _ => (),
                    };
                    

~~~